tools/filter_cells_and_genes/filter_cells_and_genes.xml

```diff
@@ -0,0 +1,246 @@
+      <requirement type="package" version="1.0.0=r3.4.1_0">bioconductor-singlecellexperiment</requirement>
+      <requirement type="package" version="1.6.0=r341_0">r-optparse</requirement>
+   </requirements>
+   <stdio>
+      <exit_code range="1:" level="fatal" description="Tool exception" />
+   </stdio>
+   <command detect_errors="exit_code"><![CDATA[
+
+Rscript $__tool_directory__/filter_cells_and_genes.R
+    -f $file
+    -c $file_header
+    -s $file_sep
+    -m $option.is_mito
+    #if $option.is_mito == "TRUE":
+        --header_mito $option.header_mito
+    #end if
+    -p $method.package
+    #if $method.package == "scater":
+        -g $method.gene_filter
+        --min_cells $method.min_cells
+    #end if
+    #if $method.package == "Seurat":
+        --min_genes $method.min_genes
+        --min_cells $method.min_cells
+    #end if
+    --output_pdf $output_plot
+    --output_matrix $output_matrix
+
+]]></command>
+   <inputs>
+      <param name="file" type="data" format="txt" label="Raw counts of expression data"/>
+      <param name="file_sep" type="select" label="Input column separator">
+         <option value="tab" selected="true">Tabs</option>
+         <option value=",">Commas</option>
+         <option value=" ">Whitespaces</option>
+         <option value=".">Dots</option>
+         <option value="_">Underscores</option>
+         <option value="-">Dashes</option>
+   </param>
+   <param name="file_header" type="select" label="Consider first line of input file as header?">
+      <option value="TRUE" selected="true">Yes</option>
+      <option value="FALSE">No</option>
+   </param>
+   <conditional name="option">
+      <param name="is_mito" type="select" label="Are there mitochondrial genes in the expression data ? Do you want to remove cells that detect more than 20% of mitochondrial genes ?">
+         <option value="FALSE" selected="true">No</option>
+         <option value="TRUE">Yes</option>
+      </param>
+      <when value="TRUE">
+         <param name="header_mito" type="text" label="Regular expression for selection of mitochondrial genes"/>
+      </when>
+      <when value="FALSE">
+      </when>
+   </conditional>
+   <conditional name="method">
+      <param name="package" type="select" label="Which package/method do you want to use ?">
+         <option value="scater" selected="true">scater</option>
+         <option value="Seurat">Seurat</option>
+      </param>
+      <when value="scater">
+         <param name="gene_filter" type="select" label="How do you want to filter out genes ?">
+            <option value="low.abundances" selected="true">low abundances genes based on library size</option>
+            <option value="min.cells">genes that aren't detected in a certain number of cells</option>
+         </param>
+         <param name="min_cells" type="integer" value="3" label="What threshold do you want to apply ?"/>
```

Comment:
Threshold is not meaningful in this context. 100 % sure that 95% users will guess instead of understand. On 50% of these will guess bad.
Plus, I am lost: the same threshold parameter is used for `not in a fraction of cells` and `low abundances genes relative to library size`. Does that mean that this is a "contextual parameter" ? I yes, this is not a good idea at all.

Reply:
Ok it was a contextual parameter, used by two different functions, I will modify that part dans separate this parameter in two.